### PR TITLE
Ensure length check on unconstrained list always passes

### LINKED ISSUE #1171: TraitListObject._validate_length should not be called for non-length-constrained lists
We're currently calling `TraitListObject._validate_length` for non-length-constrained lists, and we shouldn't be.

This is likely due to using a default minimum length of `0` rather than `None` somewhere.

### PR DESCRIPTION
**Pull request by mdickinson**
This PR is a workaround to make sure that we don't raise something that looks like a length error on an unconstrained list.

This isn't the solution I'd prefer - it's the least invasive solution that unblocks the Traits 6.1.0 release.

For Traits 6.2.0, I'd like to rework so that the `minlen` and `maxlen` attributes of a `List` trait are `None` for an unconstrained list, and have the rest of the code interpret that properly. But that feels like too invasive and risky a fix this close to the release, so for 6.1.0 we get this workaround instead.

Fixes #1171

### CODE REVIEW

**Comment by mdickinson** [comment]
Note to self: don't merge this before adding a changelog entry.

**Comment by mdickinson** [comment]
I think this fixes https://github.com/enthought/pyface/issues/508, too.

File: traits/trait_list_object.py [hunk 1]
```diff
@@ -797,7 +797,7 @@ def remove(self, value):
             If the value is not present.
         """
 
-        self._validate_length(len(self) - 1)
+        self._validate_length(max(len(self) - 1, 0))
```
Comment (kitchoi) [inline traits/trait_list_object.py:800]: What about the one in `__delitem__`? 
I was looking for all occurrences of `_validate_length(X - Y)`. I see the one in `__setitem__` probably doesn't need changing.

Comment (mdickinson) [reply traits/trait_list_object.py:800]: So the one in `__delitem__` extracts the slice first; the length of that slice is bounded by the length of the original list, so we shouldn't be able to go negative.

Comment (mdickinson) [reply traits/trait_list_object.py:800]: Ah, but not if we have a simple index rather than a slice. Good catch. Will fix.

Comment (mdickinson) [reply traits/trait_list_object.py:800]: Fixed! Thank you.

**Comment by mdickinson** [comment]
Merge conflict in the changelog. Fixing that, then merging.

**Pull request merged**